plot-builder/src/commonMain/kotlin/jetbrains/datalore/plot/builder/tooltip/layout/LayoutManager.kt

```diff
@@ -100,8 +109,31 @@ class LayoutManager(
         return placementList
     }
 
-    private fun rearrangeWithoutOverlapping(tooltips: List<PositionedTooltip>): List<PositionedTooltip> {
-        if (tooltips.isEmpty()) {
+    private fun calculateCornerTooltipsPosition(cornerTooltips: List<MeasuredTooltip>): List<PositionedTooltip> {
+        val placementList = ArrayList<PositionedTooltip>()
+
+        val tooltipsHeight = cornerTooltips.sumByDouble { it.size.y } + MARGIN_BETWEEN_TOOLTIPS * cornerTooltips.size
+        val verticalTooltipRange = when (myTooltipAnchor) {
+            TooltipAnchor.TOP_LEFT,
+            TooltipAnchor.TOP_RIGHT -> rightAligned(myVerticalSpace.start(), tooltipsHeight, MARGIN_BETWEEN_TOOLTIPS)
+            else -> leftAligned(myVerticalSpace.end(), tooltipsHeight, MARGIN_BETWEEN_TOOLTIPS)
+        }
+
+        var tooltipY = verticalTooltipRange.start()
+        cornerTooltips.forEach { tooltip ->
+            val positionedTooltip = calculatePlotCornerTooltipPosition(tooltip, tooltipY, verticalTooltipRange)
+            placementList.add(positionedTooltip)
+            tooltipY += positionedTooltip.height + MARGIN_BETWEEN_TOOLTIPS
+        }
+
+        return placementList
+    }
+
+    private fun rearrangeWithoutOverlapping(
+        tooltips: List<PositionedTooltip>,
+        fixTooltips: List<PositionedTooltip>
```

Comment:
Introduce calculated `CORNER_TOOLTIP` kind (something like extension method tooltips.selectCorner()), use in a same way as other `enum` kinds.